File: python/103/parser2.py
class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.current = 0

    def parse(self):
        return self._parse_expression()

    def _parse_expression(self):
        node = self._parse_term()
        while self._current_token() in ('+', '-'):
            operator = self._current_token()
            self._consume()
            right = self._parse_term()
            if operator == '+':
                node = ('Add', node, right)
            elif operator == '-':
                node = ('Subtract', node, right)
        return node

    def _parse_term(self):
        node = self._parse_factor()
        while self._current_token() in ('*', '/'):
            operator = self._current_token()
            self._consume()
            right = self._parse_factor()
            if operator == '*':
                node = ('Multiply', node, right)
            elif operator == '/':
                node = ('Divide', node, right)
        return node

    def _parse_factor(self):
        token = self._current_token()
        if token == '(':
            self._consume()
            node = self._parse_expression()
            self._consume()  # Consume ')'
            return node
        else:
            self._consume()
            return ('Number', int(token))

    def _current_token(self):
        return self.tokens[self.current] if self.current < len(self.tokens) else None

    def _consume(self):
        self.current += 1

# Przykład
tokens = ['2', "+", '2', "*", '2']
parser = Parser(tokens)
ast = parser.parse()
print(ast)
# ('Add', ('Number', 2), ('Multiply', ('Number', 2), ('Number', 2)))

File: vanilla-js/103/parser2.js
class Parser {
    constructor(tokens) {
        this.tokens = tokens;
        this.current = 0;
    }

    parse() {
        return this.parseExpression();
    }

    parseExpression() {
        let node = this.parseTerm();
        while (this.currentToken() === '+' || this.currentToken() === '-') {
            const operator = this.currentToken();
            this.consumeToken();
            const right = this.parseTerm();
            node = { type: operator === '+' ? 'Add' : 'Subtract', left: node, right: right };
        }
        return node;
    }

    parseTerm() {
        let node = this.parseFactor();
        while (this.currentToken() === '*' || this.currentToken() === '/') {
            const operator = this.currentToken();
            this.consumeToken();
            const right = this.parseFactor();
            node = { type: operator === '*' ? 'Multiply' : 'Divide', left: node, right: right };
        }
        return node;
    }

    parseFactor() {
        const token = this.currentToken();
        if (token === '(') {
            this.consumeToken();
            const node = this.parseExpression();
            this.consumeToken(); // Consume ')'
            return node;
        } else {
            this.consumeToken();
            return { type: 'Number', value: parseInt(token, 10) };
        }
    }

    currentToken() {
        return this.tokens[this.current];
    }

    consumeToken() {
        this.current++;
    }
}

// Przykład
const tokens = ['(', '2', '+', '2', ')', '*', '2']
const parser = new Parser(tokens);
const ast = parser.parse();
console.log(JSON.stringify(ast, null, 2));

// {
//   "type": "Multiply",
//   "left": {
//     "type": "Add",
//     "left": {
//       "type": "Number",
//       "value": 2
//     },
//     "right": {
//       "type": "Number",
//       "value": 2
//     }
//   },
//   "right": {
//     "type": "Number",
//     "value": 2
//   }
// }
